tests/test_extensions/test_ext_intersphinx.py

```diff
@@ -427,6 +430,81 @@ def test_load_mappings_fallback(tmp_path, app, status, warning):
     assert isinstance(rn, nodes.reference)
 
 
+@pytest.mark.sphinx('dummy', testroot='basic')
+def test_load_mappings_cache_update(make_app, app_params):
+    def make_invdata(i: int) -> bytes:
+        headers = f'''\
+# Sphinx inventory version 2
+# Project: foo
+# Version: {i}
+# The remainder of this file is compressed with zlib.
+'''.encode()
+        line = f'module{i} py:module 0 foo.html#module-$ -\n'.encode()
+        return headers + zlib.compress(line)
+
+    class InventoryHandler(http.server.BaseHTTPRequestHandler):
+        def do_GET(self):
+            self.send_response(200, 'OK')
+
+            if self.path.startswith('/old/'):
+                data = make_invdata(1)
+            elif self.path.startswith('/new/'):
+                data = make_invdata(2)
+            else:
+                data = b''
+
+            self.send_header('Content-Length', str(len(data)))
+            self.end_headers()
+            self.wfile.write(data)
+
+        def log_message(*args, **kwargs):
+            pass
+
+    # clean build
+    args, kwargs = app_params
+    _ = make_app(*args, freshenv=True, **kwargs)
+    _.build()
+
+    baseconfig = {'extensions': ['sphinx.ext.intersphinx']}
+    url_old = 'http://localhost:7777/old'
+    url_new = 'http://localhost:7777/new'
+
+    with http_server(InventoryHandler):
+        confoverrides1 = baseconfig | {'intersphinx_mapping': {'foo': (url_old, None)}}
+        app1 = make_app(*args, confoverrides=confoverrides1, **kwargs)
+        app1.build()
+        # the inventory when querying the 'old' URL
+        entry1 = {'module1': ('foo', '1', f'{url_old}/foo.html#module-module1', '-')}
+
+        assert list(app1.env.intersphinx_cache) == [url_old]
+        assert app1.env.intersphinx_cache[url_old][0] == 'foo'
+        assert app1.env.intersphinx_cache[url_old][2] == {'py:module': entry1}
+        assert app1.env.intersphinx_named_inventory == {'foo': {'py:module': entry1}}
+
+        # switch to new url and assert that the old URL is no more stored
+        confoverrides2 = baseconfig | {'intersphinx_mapping': {'foo': (url_new, None)}}
+        app2 = make_app(*args, confoverrides=confoverrides2, **kwargs)
+        app2.build()
+        entry2 = {'module2': ('foo', '2', f'{url_new}/foo.html#module-module2', '-')}
+
+        assert list(app2.env.intersphinx_cache) == [url_new]
+        assert app2.env.intersphinx_cache[url_new][0] == 'foo'
+        assert app2.env.intersphinx_cache[url_new][2] == {'py:module': entry2}
+        assert app2.env.intersphinx_named_inventory == {'foo': {'py:module': entry2}}
+
+        # switch back to old url
+        confoverrides3 = baseconfig | {'intersphinx_mapping': {'foo': (url_old, None)}}
+        app3 = make_app(*args, confoverrides=confoverrides3, **kwargs)
+        app3.build()
+        # same as entry 1
+        entry3 = {'module1': ('foo', '1', f'{url_old}/foo.html#module-module1', '-')}
```

Comment:
Optional: we could add some cognitive distancing between the `objects.inv` file-format number (1/2) and the project/module version number.

```suggestion
    def make_invdata(project_name: str = 'foo', project_version: int) -> bytes:
        headers = f'''\
# Sphinx inventory version 2
# Project: {project_name}
# Version: {project_version}
# The remainder of this file is compressed with zlib.
'''.encode()
        line = f'module{project_version} py:module 0 foo.html#module-$ -\n'.encode()
        return headers + zlib.compress(line)

    class InventoryHandler(http.server.BaseHTTPRequestHandler):
        def do_GET(self):
            self.send_response(200, 'OK')

            if self.path.startswith('/old/'):
                data = make_invdata(project_version=99)
            elif self.path.startswith('/new/'):
                data = make_invdata(project_version=100)
            else:
                data = b''

            self.send_header('Content-Length', str(len(data)))
            self.end_headers()
            self.wfile.write(data)

        def log_message(*args, **kwargs):
            pass

    # clean build
    args, kwargs = app_params
    _ = make_app(*args, freshenv=True, **kwargs)
    _.build()

    baseconfig = {'extensions': ['sphinx.ext.intersphinx']}
    url_old = 'http://localhost:7777/old'
    url_new = 'http://localhost:7777/new'

    with http_server(InventoryHandler):
        confoverrides1 = baseconfig | {'intersphinx_mapping': {'foo': (url_old, None)}}
        app1 = make_app(*args, confoverrides=confoverrides1, **kwargs)
        app1.build()
        # the inventory when querying the 'old' URL
        entry1 = {'module99': ('foo', '99', f'{url_old}/foo.html#module-module99', '-')}

        assert list(app1.env.intersphinx_cache) == [url_old]
        assert app1.env.intersphinx_cache[url_old][0] == 'foo'
        assert app1.env.intersphinx_cache[url_old][2] == {'py:module': entry1}
        assert app1.env.intersphinx_named_inventory == {'foo': {'py:module': entry1}}

        # switch to new url and assert that the old URL is no more stored
        confoverrides2 = baseconfig | {'intersphinx_mapping': {'foo': (url_new, None)}}
        app2 = make_app(*args, confoverrides=confoverrides2, **kwargs)
        app2.build()
        entry2 = {'module100': ('foo', '100', f'{url_new}/foo.html#module-module100', '-')}

        assert list(app2.env.intersphinx_cache) == [url_new]
        assert app2.env.intersphinx_cache[url_new][0] == 'foo'
        assert app2.env.intersphinx_cache[url_new][2] == {'py:module': entry2}
        assert app2.env.intersphinx_named_inventory == {'foo': {'py:module': entry2}}

        # switch back to old url
        confoverrides3 = baseconfig | {'intersphinx_mapping': {'foo': (url_old, None)}}
        app3 = make_app(*args, confoverrides=confoverrides3, **kwargs)
        app3.build()
        # same as entry 1
        entry3 = {'module99': ('foo', '99', f'{url_old}/foo.html#module-module99', '-')}
```

Reply:
I refactored the test so that in the future we can easily generate those inventories as well without having too many variable names to change.